libkbfs/disk_block_cache.go

```diff
@@ -0,0 +1,242 @@
+package libkbfs
+
+import (
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/logger"
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	"github.com/keybase/kbfs/tlf"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultDiskBlockCacheMaxBytes uint64 = 10 * (1 << 30)
+	evictionConsiderationFactor   uint   = 3
+	blockDbFilename               string = "diskCacheBlocks.leveldb"
+	lruDbFilename                 string = "diskCacheLRU.leveldb"
+)
+
+type diskBlockCacheEntry struct {
+	Buf        []byte
+	ServerHalf kbfscrypto.BlockCryptKeyServerHalf
+}
+
+type diskBlockCacheConfig interface {
+	codecGetter
+	logMaker
+}
+
+// DiskBlockCacheStandard is the standard implementation for DiskBlockCache.
+type DiskBlockCacheStandard struct {
+	config   diskBlockCacheConfig
+	maxBytes uint64
+	log      logger.Logger
+	// protects everything below
+	lock     sync.RWMutex
+	isClosed bool
+	blockDb  *leveldb.DB
+	lruDb    *leveldb.DB
+}
+
+var _ DiskBlockCache = (*DiskBlockCacheStandard)(nil)
+
+func newDiskBlockCacheStandard(config diskBlockCacheConfig, dirPath string,
+	maxBytes uint64) (*DiskBlockCacheStandard, error) {
+	log := config.MakeLogger("KBC")
+	blockDbPath := filepath.Join(dirPath, blockDbFilename)
+	blockDb, err := leveldb.OpenFile(blockDbPath, leveldbOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	lruDbPath := filepath.Join(dirPath, lruDbFilename)
+	lruDb, err := leveldb.OpenFile(lruDbPath, leveldbOptions)
+	if err != nil {
+		blockDb.Close()
+		return nil, err
+	}
+	return &DiskBlockCacheStandard{
+		config:   config,
+		maxBytes: maxBytes,
+		log:      log,
+		blockDb:  blockDb,
+		lruDb:    lruDb,
+	}, nil
+}
+
+// TODO: Fix getSizes(), where leveldb.DB.SizeOf() currently doesn't work for
+// the full range.
+func (cache *DiskBlockCacheStandard) getSizes() (blockSize int64, lruSize int64, err error) {
+	blockSizes, err := cache.blockDb.SizeOf([]util.Range{{}})
+	if err != nil {
+		return -1, -1, err
+	}
+	lruSizes, err := cache.lruDb.SizeOf([]util.Range{{}})
+	if err != nil {
+		return -1, -1, err
+	}
+	return blockSizes[0], lruSizes[0], nil
+}
+
+func (cache *DiskBlockCacheStandard) compactCaches(ctx context.Context) {
+	cache.blockDb.CompactRange(util.Range{})
+	cache.lruDb.CompactRange(util.Range{})
+	bSize, lruSize, err := cache.getSizes()
+	cache.log.CDebugf(ctx, "Disk Cache bSize=%d lruSize=%d err=%+v", bSize, lruSize, err)
+}
+
+func (cache *DiskBlockCacheStandard) updateLruLocked(tlfID tlf.ID,
+	blockBytes []byte) error {
+	key := append(tlfID.Bytes(), blockBytes...)
+	val, err := time.Now().MarshalBinary()
+	if err != nil {
+		return err
+	}
+	cache.lruDb.Put(key, val, nil)
```

Comment:
Since `lock` is only read-locked where this is called, it's possible to have a race where the lru time jumps backward for a given block that's accessed twice concurrently.  Not a big deal at all, not even sure it's worth a comment, just pointing it out.